coded_tools/deep_rag/create_networks.py

```diff
@@ -0,0 +1,360 @@
+        string_replacements: Dict[str, Any] = {
+            "one_group": group.get("name"),
+            "group_description": group.get("description"),
+            "structure_description": self.grouping_json.get("description"),
+            "title": self.grouping_json.get("name"),
+        }
+        front_man = self.filter_agent(front_man, string_replacements)
+
+        front_man["tools"] = tools
+
+        return front_man
+
+    async def reserve_leaf_networks(self, reservationist: Reservationist,
+                                    name_to_network: Dict[str, Dict[str, Any]]) \
+            -> Dict[Reservation, Dict[str, Any]]:
+        """
+        Creates reservations for each named network
+        """
+        deployments: Dict[Reservation, Dict[str, Any]] = {}
+
+        for name, network in name_to_network.items():
+            filtered_name: str = name.replace(" ", "_")
+            reservation: Reservation = await reservationist.reserve(lifetime_in_seconds=self.LIFETIME,
+                                                                    prefix=filtered_name)
+            deployments[reservation] = network
+
+        return deployments
+
+    def make_group_network(self, reservations: List[Reservation]) -> Dict[str, Any]:
+        """
+        Creates a final front-man network for the rest.
+        """
+
+        agent_spec: Dict[str, Any] = deepcopy(self.network_template)
+        tools: List[Dict[str, Any]] = agent_spec.get("tools")
+
+        # We don't need the content node, we are using external networks for those.
+        _ = tools.pop()
+
+        # Make a list of the external networks for the reservations to reference as tools
+        external_tools: List[str] = []
+        for reservation in reservations:
+            res_id: str = reservation.get_url()
+            if not res_id.startswith("/") or not res_id.startswith("http"):
+                res_id = "http://localhost/" + res_id
+            external_tools.append(res_id)
+
+        # Start out with the front man from the template, but replace him with what's made.
+        front_man: Dict[str, Any] = tools[self.TEMPLATE_FRONT_MAN_INDEX]
+        front_man = self.create_front_man(front_man, self.grouping_json, external_tools)
+        front_man["function"]["description"] = front_man["user_prompt"]
+        tools[self.TEMPLATE_FRONT_MAN_INDEX] = front_man
+
+        return agent_spec
+
+    def assemble_reservation_info(self, reservations: List[Reservation]) -> List[Dict[str, Any]]:
+        """
+        Assemble the list of networks available to the user.
+        """
+        reservation_info: List[Dict[str, Any]] = []
+        for reservation in reservations:
+            one_info: Dict[str, Any] = {
+                "reservation_id": reservation.get_reservation_id(),
+                "lifetime_in_seconds": reservation.get_lifetime_in_seconds(),
+                "expiration_time_in_seconds": reservation.get_expiration_time_in_seconds(),
+            }
```

Comment:
This is the same format that copy_cat uses.
Idea is to have a convention so clients know what to expect generically.